### PR TITLE
fix: crash on print cancellation and silent print settings

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Fixes https://github.com/electron/electron/issues/19122.
Fixes https://github.com/electron/electron/issues/17077.
Fixes https://github.com/electron/electron/issues/19108.
Fixes https://github.com/electron/electron/issues/19474.

In an earlier PR, I introduced the ability to set custom print options in `webContents.print()`.

However, this meant that the `base::DictionaryValue` passed in with these settings would never be `null`, because if even one option is set than they _all_ have to be set with a default to prevent a DCHECK that gets hit if [this function](https://cs.chromium.org/chromium/src/printing/print_settings_conversion.cc?type=cs&q=printsettingsfromjob&g=0&l=78) returns false. As a result, only the `PrintHostMsg_UpdatePrintSettings` [IPC message](https://github.com/electron/electron/blob/4bea46cccdc8f8ed0dd9c015dc8b496a3fc9a268/patches/chromium/printing.patch#L504-L505) would be sent, and printing would silently fail as no defaults were properly initialized. More specifically, since no defaults were set, when a new `PrinterQuery` was created it would have no idea about previously existing or defaulted settings, since settings were reset at the beginning of print updates.

This fixes that by changing codepaths to _intentionally_ only send the IPC message to update print settings (`PrintHostMsg_UpdatePrintSettings`), and then each time pursuing the following codepath: 
1) Reset settings from previous job
2) Call `printing_context_->UseDefaultSettings()` in order to set the `settings_` member with default settings
3) Call `printing_context_->UpdatePrintSettings()` with our new settings to only change the options that were passed in.

These custom settings should work now for both silent and non-silent printing: this can be seeing by calling:

```
win.webContents.print({ copies: 3 }, (success, reason) => {
    console.log(`printing ${success ? 'did' : 'did not'} succeed: ${reason}`)
})
```

to see that '3' is automatically placed in the UI settings of the native print dialog.

**Secondary Fix:**

Due to changes in the lifetime of PrinterQuery, a crash occurred if we posted a task to notify of print cancellation and did _not_ also call the callback. This removes that gating and now will call the callback regardless of cancellation status.

I chose not to put this secondary fix in a standalone PR because the primary issue above meant that it isn't even being surfaced to users as it causes a silent failure before a UI printing dialog is ever shown.

cc @jkleinsc @deepak1556 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fixed a crash on manual print cancellation as well as an issue with `deviceName` not working.

### CODE REVIEW

**Comment by fleabitten232** [comment]
I'm currently using your changes in a small, heavily print dependent project that needs to silently print to both a standard A4 size/laser printer combination and a notoriously temperamental 6x4inch thermal label printer. I'm happy to share my experiences if you would find that helpful -- just let me know where if here isn't appropriate. 

I've now mostly finished moving from electron@latest to electron@beta which broke a label design that was previously working. In latest, it would correctly print landscape (albeit not silently). 

In the beta (without any code/css changes), the same label was adamant that it wanted to be portrait. It would print massively scaled down so that the landscape design would fit to the width of a portrait page. This happened with 'silent' sent to both true and false.

I couldn't immediately fix the problem through any print dialog settings and unfortunately I ran out of labels before I could explore deeper. The last thing I noticed was that setting the page to 'landscape' would actually rotate it 180 degrees, rather than 90. Although the printer does have a setting to rotate labels 180 degrees, I'm wondering if the page is somehow being 'landscaped' twice.

I'll have some more labels in the next 24 hours so I'll keep you updated. Unfortunately the project has a looming deadline so if I can't get it working soon I'll have to just rotate the label design 90deg using CSS as a hack and call it done.

File: spec/ts-smoke/electron/main.ts [hunk 1]
```diff
@@ -83,7 +83,7 @@ app.on('ready', () => {
   mainWindow.webContents.setVisualZoomLevelLimits(50, 200)
   mainWindow.webContents.setLayoutZoomLevelLimits(50, 200)
 
-  mainWindow.webContents.print({ silent: true, printBackground: false })
+  mainWindow.webContents.print({ silent: true })
```
Comment (jkleinsc) [inline spec/ts-smoke/electron/main.ts]: Is this change necessary?

**Comment by codebytere** [comment]
@deepak1556 i've managed to fix the underlying issue, and have updated the PR body accordingly :)

File: spec/ts-smoke/electron/main.ts (hunk 1 above)
Comment (codebytere) [reply spec/ts-smoke/electron/main.ts]: nope, removed with latest changes!

File: patches/chromium/printing.patch [hunk 2]
```diff
@@ -576,3 +580,55 @@ index 71c0c15217b62cd7a6087c6d9ae50481f9041d5f..e2250a66517dbd909cd3b5407173ad91
...
+diff --git a/printing/print_settings_conversion.cc b/printing/print_settings_conversion.cc
+index 2563ae6a87b2354ff2f2b45c17f61d2f44910efa..1f34f905791f38a44ae8c892cc9cfb38d15b659d 100644
+--- a/printing/print_settings_conversion.cc
++++ b/printing/print_settings_conversion.cc
+@@ -190,11 +190,13 @@ bool PrintSettingsFromJobSettings(const base::Value& job_settings,
+ 
+   settings->set_dpi_xy(dpi_horizontal.value(), dpi_vertical.value());
+ #endif
++  std::string print_device_name = *device_name;
++  if (print_device_name != "")
```
Comment (deepak1556) [inline patches/chromium/printing.patch]: Can be `!device_name->empty()`

File: shell/browser/api/atom_api_web_contents.cc [hunk 3]
```diff
@@ -1643,6 +1651,8 @@ void WebContents::Print(mate::Arguments* args) {
   options.Get("landscape", &landscape);
   settings.SetBoolean(printing::kSettingLandscape, landscape);
 
+  // We set the default to empty string here and only update
+  // if at the Chromium level if it's non-empty
```
Comment (deepak1556) [inline shell/browser/api/atom_api_web_contents.cc]: What happens when an empty device name is passed along ?

File: patches/chromium/printing.patch [hunk 4]
```diff
@@ -496,24 +509,17 @@ index ef580254bd8feba84ac02924b77b9b4feaf14d96..57a8264a7174b440ed1d6bfe9c2c3e21
...
-+    bool canceled = false;
-+    Send(new PrintHostMsg_UpdatePrintSettings(
-+        routing_id(), cookie, new_settings, &settings, &canceled));
-+    if (canceled)
-+      return false;
-+  }
++  // new_settings will never be empty, always send the update IPC message
++  bool canceled = false;
++  Send(new PrintHostMsg_UpdatePrintSettings(
++      routing_id(), 0, new_settings, &settings, &canceled));
```
Comment (deepak1556) [inline patches/chromium/printing.patch]: So by setting the cookie value to be zero we create a new PrinterQuery every time instead of reusing one if exists for same settings. Any reason this is the case ?

Comment (codebytere) [reply patches/chromium/printing.patch]: @deepak1556 this was codification of implicit behavior; becuase we set the defaults ourselves the params member would never been initialized, so that ternary would always have evaluated to 0 (i did some testing to ensure this was the case)

File: shell/browser/api/atom_api_web_contents.cc (hunk 3 above)
Comment (codebytere) [reply shell/browser/api/atom_api_web_contents.cc]: https://github.com/electron/electron/pull/19598#discussion_r310895212 that's what is checked here; if no value is passed by the user it is set to the value initialized in `printing_context_->UseDefaultValues()` which is the system's default printer.

**Comment by codebytere** [comment]
CI failures are unrelated and will be fixed by: https://github.com/electron/electron/pull/19652

File: shell/browser/api/atom_api_web_contents.cc (hunk 3 above)
Comment (deepak1556) [reply shell/browser/api/atom_api_web_contents.cc]: Sorry for being unclear, I meant what happens when we pass along empty device string, does it still pick the default value or crash ? I am trying to understand if its necessary for an empty device name check.

Comment (codebytere) [reply shell/browser/api/atom_api_web_contents.cc]: it errors if we don't do this, since it sets the printer to empty string which necessarily won't exist!

Comment (deepak1556) [reply shell/browser/api/atom_api_web_contents.cc]: Got it, thanks for clarifying.
